framework/src/functions/PiecewiseConstantFromCSV.C

```diff
@@ -0,0 +1,115 @@
+  // but it's also safer to have that parameter be close to the use case, and not only in the UO
+  params.addRequiredParam<MooseEnum>("read_type",
+                                     MooseEnum("element voronoi block node"),
+                                     "Organization of data in the CSV file: "
+                                     "element:by element "
+                                     "node: by node "
+                                     "voronoi:nearest neighbor / voronoi tesselation structure "
+                                     "block:by mesh block");
+  // We need one ghost layer in case we need to retrieve the id or subdomain id from an
+  // element that is on the other side of a domain boundary
+  params.set<unsigned short>("ghost_layers") = 1;
+
+  params.addClassDescription("Uses data read from CSV to assign values");
+  return params;
+}
+
+PiecewiseConstantFromCSV::PiecewiseConstantFromCSV(const InputParameters & parameters)
+  : Function(parameters),
+    _read_prop_user_object(nullptr),
+    _column_number(getParam<unsigned int>("column_number")),
+    _read_type(getParam<MooseEnum>("read_type").getEnum<ReadTypeEnum>()),
+    _point_locator(_ti_feproblem.mesh().getPointLocator())
+{
+  if (_column_number < _ti_feproblem.mesh().dimension() && _read_type == ReadTypeEnum::VORONOI)
+    mooseWarning(
+        "The column requested in the function is likely to just be containing point coordinates");
+}
+
+void
+PiecewiseConstantFromCSV::initialSetup()
+{
+  // Get a pointer to the ElementPropertyReadFile. A pointer is used because the UserObject is not
+  // available during the construction of the function
+  _read_prop_user_object = &getUserObject<PropertyReadFile>("read_prop_user_object");
+
+  if (_read_type != _read_prop_user_object->getReadType())
+    paramError("read_type", "The PropertyReadFile UO should have the same read_type parameter.");
+}
+
+Real
+PiecewiseConstantFromCSV::value(Real, const Point & p) const
+{
+  if (_read_type == ReadTypeEnum::ELEMENT || _read_type == ReadTypeEnum::BLOCK)
+  {
+    // This is somewhat inefficient, but it allows us to retrieve the data in the
+    // CSV by element or by block.
+    std::set<const Elem *> candidate_elements;
+    (*_point_locator)(p, candidate_elements);
+
+    // Find the element with the lowest ID (that is also on the domain ???????)
+    unsigned int min_elem_id = std::numeric_limits<unsigned int>::max();
+    const Elem * min_id_elem;
+    unsigned int num_candidates = 0;
+
+    std::set<const Elem *>::iterator it = candidate_elements.begin();
+    while (it != candidate_elements.end())
+    {
+      if ((*it)->id() < min_elem_id)
+      {
+        min_elem_id = (*it)->id();
+        min_id_elem = (*it);
+        num_candidates++;
+      }
+      it++;
+    }
+
+    if (num_candidates == 0)
+      mooseError("No element located at ", p, " to search in element or block sorted CSV values");
```

Comment:
```suggestion
    const Elem * min_id_elem = nullptr;
    for (const auto & elem : candidate_elements)
      if (!elem || elem->id() < min_id_elem->id())
        min_id_elem = elem;
    if (!min_id_elem)
      mooseError("No element located at ", p, " to search in element or block sorted CSV values");
```

This isn't great but I see why you do it